### PR TITLE
JIT: Improve consistency around indir flags; directly copy flags in `gtCloneExpr`

### LINKED ISSUE #97524: JIT: Assertion failed '!"Extra flags on tree"' during 'Post-Morph'
### Error Blob

```json
{
  "ErrorMessage": "",
  "BuildRetry": false,
  "ErrorPattern": "Extra flags on tree",
  "ExcludeConsoleLog": false
}
```

### Reproduction Steps

Example test run: https://dev.azure.com/dnceng-public/public/_build/results?buildId=540840&view=results
Example console log: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-97488-merge-c4d5c6c8a7f64d9baf/System.Runtime.InteropServices.Tests/1/console.3fbb3808.log?helixlogtype=result

 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=540840
**Error message validated:** `Extra flags on tree`
**Result validation:** :white_check_mark: Known issue matched with the provided build.
**Validation performed at:** 1/25/2024 9:05:04 PM UTC

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[541598](https://dev.azure.com/dnceng-public/public/_build/results?buildId=541598)|dotnet/runtime|[System.Runtime.InteropServices.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=541598&view=ms.vss-test-web.build-test-results-tab&runId=12830290&resultId=212118)||
|[540840](https://dev.azure.com/dnceng-public/public/_build/results?buildId=540840)|dotnet/runtime|[System.Runtime.InteropServices.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=540840&view=ms.vss-test-web.build-test-results-tab&runId=12809492&resultId=212059)|dotnet/runtime#97488|
|[540774](https://dev.azure.com/dnceng-public/public/_build/results?buildId=540774)|dotnet/runtime|[System.Runtime.InteropServices.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=540774&view=ms.vss-test-web.build-test-results-tab&runId=12807304&resultId=212110)|dotnet/runtime#97517|
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|3|3|3|

### PR DESCRIPTION
**Pull request by jakobbotsch**
- Switch `GT_STORE_DYN_BLK` to be on the same plan as other indirs; add a `gtNewStoreDynBlkNode` and use `gtInitializeIndirNode` to set its flags correctly. Add support for it to `GenTree::SetIndirExceptionFlags`. Remove conservative flag setting from its constructor.
- Fix propagation of `GTF_EXCEPT` from `GT_CMPXCHG`'s data source inside `GenTree::SetIndirExceptionFlags`
- Set flags directly from the source node in `gtCloneExpr`

Fix #97524

### CODE REVIEW

**Review by SingleAccretion — APPROVED** [review]

File: src/coreclr/jit/gentree.cpp [hunk 1]
```diff
@@ -8603,6 +8603,34 @@ GenTreeBlk* Compiler::gtNewStoreBlkNode(ClassLayout* layout, GenTree* addr, GenT
...
+GenTreeStoreDynBlk* Compiler::gtNewStoreDynBlkNode(GenTree*     addr,
+                                                   GenTree*     data,
+                                                   GenTree*     dynamicSize,
+                                                   GenTreeFlags indirFlags)
+{
+    assert((indirFlags & GTF_IND_INVARIANT) == 0);
+    assert(data->IsInitVal() || data->OperIs(GT_IND));
+
+    GenTreeStoreDynBlk* store = new (this, GT_STORE_DYN_BLK) GenTreeStoreDynBlk(addr, data, dynamicSize);
+    store->gtFlags |= GTF_ASG;
```
Comment (EgorBo) [inline src/coreclr/jit/gentree.cpp:8627]: nit: it's already set in GenTreeStoreDynBlk ctror

**Review by EgorBo — APPROVED** [review]

File: src/coreclr/jit/gentree.cpp (hunk 1 above)
Comment (jakobbotsch) [reply src/coreclr/jit/gentree.cpp:8627]: I removed that in this PR (just to make it consistent with `gtNewStoreBlkNode`)

Comment (EgorBo) [reply src/coreclr/jit/gentree.cpp:8627]: ah, I am blind

**Pull request merged**